File: 2020/8.py
#!/usr/bin/env python

class Computer(object):
  def __init__(self):
    self.accumulator = 0
    self.code = []
    self.original_code = []

  def load_code(self, lines):
    self.accumulator = 0
    for line in lines:
      op, count = line.split()
      if count.startswith("+"):
        number = int(count[1:])
      else:
        number = int(count)
      self.code.append((op, number))
    self.original_code = list(self.code)

  def run(self):
    i = 0
    seen = set()
    while True:
      if i >= len(self.code):
        print("Normal completion! Accumulator is %d" % self.accumulator)
        return True
      if i in seen:
        print("Loop detected. Accumulator is %d" % self.accumulator)
        return False
      seen.add(i)
      op, arg = self.code[i]
      if op == "acc":
        self.accumulator += arg
        i += 1
      elif op == "jmp":
        i += arg
      elif op == "nop":
        i += 1

  def mutate_and_run(self):
    i = 0
    while True:
      self.accumulator = 0
      modified = False
      if self.code[i][0] == "jmp":
        self.code[i] = ("nop", self.code[i][1])
        modified = True
      elif self.code[i][0] == "nop":
        self.code[i] = ("jmp", self.code[i][1])
        modified = True
      i += 1

      if modified:
        ok = self.run()
        if ok:
          print "Yay we're done"
          exit()
        else:
          print "RESET"
          self.code = list(self.original_code)

lines = [
  "nop +0",
  "acc +1",
  "jmp +4",
  "acc +3",
  "jmp -3",
  "acc -99",
  "acc +1",
  "jmp -4",
  "acc +6",
]

inputfile = "input8.txt"
with open(inputfile, "r") as f:
  lines = [x.strip() for x in f.readlines()]

computer = Computer()
computer.load_code(lines)
result = computer.run()

computer.mutate_and_run()

File: 2020/8.kt
// Advent of code 2020 day 8
import java.io.File
import java.lang.*

class Computer() {
  var accumulator = 0
  var code = mutableListOf<Pair<String, Int>>()
  var original_code = listOf<Pair<String, Int>>()

  fun load_code(lines: MutableList<String>) {
    this.accumulator = 0
    for (line in lines) {
      var pieces = line.split(" ")
      var op = pieces[0]
      var count = pieces[1]
      var number:Int
      if (count.startsWith("+")) {
        number = count.substring(1).toInt()
      } else {
        number = count.toInt()
      }
      this.code.add(Pair(op, number))
    }
    this.original_code = this.code.toList()
  }

  fun run(): Boolean {
    var i = 0
    var seen = mutableSetOf<Int>()
    while (true) {
      if (i >= this.code.size) {
        println("Normal completion! Accumulator is ${this.accumulator}")
        return true
      }
      if (i in seen) {
        println("Loop detected. Accumulator is ${this.accumulator}")
        return false
      }
      seen.add(i)
      val op = this.code[i].first
      val arg = this.code[i].second
      if (op == "acc") {
        this.accumulator += arg
        i += 1
      } else if (op == "jmp") {
        i += arg
      } else if (op == "nop") {
        i += 1
      }
    }
  }

  fun mutate_and_run() {
    var i = 0
    while (true) {
      this.accumulator = 0
      var modified = false
      if (this.code[i].first == "jmp") {
        this.code[i] = Pair<String, Int>("nop", this.code[i].second)
        modified = true
      } else if (this.code[i].first == "nop") {
        this.code[i] = Pair<String, Int>("jmp", this.code[i].second)
        modified = true
      }
      i += 1

      if (modified) {
        val ok = this.run()
        if (ok) {
          println("Yay we're done")
          return
        } else {
          println("RESET")
          this.code = this.original_code.toMutableList()
        }
      }
    }
  }
}


fun main() {
  val fileName = "input8.txt"
  val lines: MutableList<String> = File(fileName).readLines().toMutableList()

  val computer = Computer()
  computer.load_code(lines)

  // Part 1
  computer.run()

  // Part 2
  computer.mutate_and_run()
}
